bin/tx-prover/src/proxy/mod.rs

```diff
@@ -40,111 +44,132 @@ impl LoadBalancer {
-                worker_queue.remove(0);
-            }
-        }
+/// Rate limiter
+static RATE_LIMITER: Lazy<Rate> = Lazy::new(|| Rate::new(Duration::from_secs(1)));
+
+// Request queue
+// ================================================================================================
+
+/// Request queue holds the list of requests that are waiting to be processed by the workers.
+/// It is used to keep track of the order of the requests to then assign them to the workers.
+pub struct RequestQueue {
+    queue: RwLock<VecDeque<u64>>,
+}
+
+impl RequestQueue {
+    /// Create a new empty request queue
+    pub fn new() -> Self {
+        Self { queue: RwLock::new(VecDeque::new()) }
+    }
+
+    /// Get the length of the queue
+    pub async fn len(&self) -> usize {
+        self.queue.read().await.len()
+    }
+
+    /// Enqueue a request
+    pub async fn enqueue(&self, request_id: u64) {
+        let mut queue = self.queue.write().await;
+        queue.push_back(request_id);
+    }
+
+    /// Dequeue a request
+    pub async fn dequeue(&self) -> Option<u64> {
+        let mut queue = self.queue.write().await;
+        queue.pop_front()
     }
 
-    /// Get the request ID from the session headers
-    pub fn get_request_id(session: &Session) -> Result<&str> {
-        session
-            .get_header("X-Request-ID")
-            .ok_or(Error::new(ErrorType::InternalError))?
-            .to_str()
-            .map_err(|_| Error::new(ErrorType::InternalError))
+    /// Peek at the first request in the queue
+    pub async fn peek(&self) -> Option<u64> {
+        let queue = self.queue.read().await;
+        queue.front().copied()
     }
 }
 
-/// Rate limiter
-static RATE_LIMITER: Lazy<Rate> = Lazy::new(|| Rate::new(Duration::from_secs(1)));
+/// Shared state. It keeps track of the order of the requests to then assign them to the workers.
+static QUEUE: Lazy<RequestQueue> = Lazy::new(RequestQueue::new);
 
-/// Shared state. It is a map of workers to a vector of request IDs
-static QUEUES: Lazy<RwLock<HashMap<Backend, Vec<String>>>> =
-    Lazy::new(|| RwLock::new(HashMap::new()));
+// RequestContext
+// ================================================================================================
 
 /// Custom context for the request/response lifecycle
-/// We use this context to keep track of the number of tries for a request.
-pub struct TriesCounter {
+/// We use this context to keep track of the number of tries for a request, the unique ID for the
+/// request, and the worker that will process the request.
+pub struct RequestContext {
+    /// Number of tries for the request
     tries: usize,
+    /// Unique ID for the request
+    request_id: u64,
+    /// Worker that will process the request
+    worker: Option<Backend>,
+}
+
+impl RequestContext {
+    /// Create a new request context
+    fn new() -> Self {
+        Self {
+            tries: 0,
+            request_id: rand::random::<u64>(),
```

Comment:
Previously, we were getting request ID from the session. Is there a reason not to do it now?

Reply:
Makes sense. I think it is probably fine to keep as is, but could we increase the value to `u128` (to make the probability of collisions truly negligible).

Reply:
A very common approach would be to use [uuid](https://crates.io/crates/uuid) here, but anything of that size should work fine